File: ios/MissionControl/Pairing.swift
import Foundation

/// Payload encoded in the pairing QR / deep link:
/// `remy://configure?url=<server>&token=<token>`
/// Older `missioncontrol://` links still pair.
struct PairingConfig {
    let url: String
    let token: String

    init(url: String, token: String) {
        self.url = url
        self.token = token
    }

    init?(from url: URL) {
        guard let scheme = url.scheme, ["remy", "missioncontrol"].contains(scheme),
              url.host == "configure",
              let items = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems,
              let server = items.first(where: { $0.name == "url" })?.value, !server.isEmpty,
              let token = items.first(where: { $0.name == "token" })?.value, !token.isEmpty else {
            return nil
        }
        self.url = server
        self.token = token
    }

    init?(fromString string: String) {
        guard let url = URL(string: string) else { return nil }
        self.init(from: url)
    }

    var pairingURL: URL? {
        var components = URLComponents()
        components.scheme = "remy"
        components.host = "configure"
        components.queryItems = [
            URLQueryItem(name: "url", value: url),
            URLQueryItem(name: "token", value: token)
        ]
        return components.url
    }

    var pairingLink: String {
        pairingURL?.absoluteString ?? ""
    }
}

extension Server {
    var pairingLink: String {
        PairingConfig(url: url, token: token).pairingLink
    }
}

File: mobile/src/lib/pairing.ts
/// Payload encoded in the pairing QR / deep link:
/// `remy://configure?url=<server>&token=<token>`
/// Older `missioncontrol://` links still pair.

export function parsePairingLink(raw: string): { url: string; token: string } | undefined {
  const trimmed = raw.trim();
  if (!trimmed) return undefined;
  try {
    const parsed = new URL(trimmed);
    if (
      (parsed.protocol !== "remy:" && parsed.protocol !== "missioncontrol:") ||
      parsed.hostname !== "configure"
    ) {
      return undefined;
    }
    const url = parsed.searchParams.get("url")?.trim();
    const token = parsed.searchParams.get("token")?.trim();
    if (!url || !token) return undefined;
    return { url, token };
  } catch {
    return undefined;
  }
}

export function formatPairCode(code: string): string {
  const digits = code.replace(/\D/g, "");
  return digits.length === 6 ? `${digits.slice(0, 3)} ${digits.slice(3)}` : digits;
}

export function hostLabel(url: string): string {
  try {
    return new URL(url).host || url;
  } catch {
    return url;
  }
}

export function threadIdFromLink(raw: string): string | undefined {
  const match = /^(?:remy|missioncontrol):\/\/chat\/(.+)$/.exec(raw.trim());
  if (!match?.[1]) return undefined;
  try {
    return decodeURIComponent(match[1]).trim() || undefined;
  } catch {
    return match[1].trim() || undefined;
  }
}
